src/dosext/mfs/mfs.c

```diff
@@ -509,133 +518,94 @@ select_drive(struct vm86_regs *state, int *drive)
   case DELETE_FILE:		/* 0x13 */
   case CREATE_TRUNCATE_FILE:	/* 0x17 */
   case FIND_FIRST:		/* 0x1b */
-    check_cds = TRUE;
+    {
+      char *fn1 = sda_filename1(sda);
```

Comment:
So why this code doesn't do
`if (fn1[1] == ':')`
so that it can just be merged with
the _NO_CDS block?

Reply:
Maybe we can. I dropped the patch that commonised the drive extraction from filename logic which avoided the need to look at it. I'll have to check each function in RBIL, but at the moment it looks like 'belt and braces' to try to get filename then CDS from SDA. I also think the logic around SET_CURRENT_DIRECTORY looks like support for the UNC drive on the fly, but specifically to disallow UNC in that function.

Reply:
I looked up all functions that use CDS path
in `select_drive()`, and they all are specified
as to take the fully qualified file name. So I
am really surprised they do not even try to
get the drive from file name.

Reply:
Yes, that seems strange.
More questions:
- If user does `DIR \\LINUX\FS\tmp` what CDS could DOS supply?
- Why does this code deal with printers, either LPTx or \\LINUX\PRN\?

It seems to me that CDS is redundant in these functions.

Reply:
> If user does DIR \\LINUX\FS\tmp what CDS could DOS supply?

No idea. I think there was some "Drive Z"
functionality in dosemu1 but I broke it. It
was used for "rescue boot" and for that
strange UNC support, and likely had a CDS
for drive Z.
Now I think we just need to remove
the remaining traces of this all. dosemu2 has the
different "rescue boot" technique, namely, just
booting from D, and if it is not available (user
messed up the setup) - it doesn't boot. Actually
it scans all drives for something bootable, so it
can find also drive F with freedos and boot from
there, so the user have to be really desperate to
break the boot in dosemu2. But no rescue drives
created on fly - this is an overkill, lets allow the
user to break his setup if he really wants to.
Also dosemu2 should respect security and not
allow to escape the sandbox, so this UNC thingy
does not fit (and it makes mfs.c too dirty).

> Why does this code deal with printers, either LPTx or \LINUX\PRN?

Could you clarify the question?
`check_sda_ffn` works with UNC PRN and
`check_dssi_fn` work with LPTx.
If you ask about the first case - this is because
`QUALIFY_FILENAME` is implemented for printers,
so their pathes are in UNC. Hope now you understand
why I wanted to implement it also for files - it is a bit
inconsistent to have printers in UNC and files in DOS
format.

Reply:
> Could you clarify the question?

Why and how does `select_drive()` establish a drive association with a printer? Printers are printers and drives are drives, why does dosemu's redirector need a drive number to process printer commands? Does DOS use generic redirector file ops `open()` or `close()` etc on printers?

BTW I never used printers under dosemu, neither `LPTx` nor `\\LINUX\PRN\?`

Reply:
> Why and how does select_drive() establish a drive association with a printer?

A hack really. `select_drive()` didn't have
`NO_DRIVE_ASSOCIATED`, and the code
looks into `drives[dd].root` - why, because
`QUALIFY_FILENAME` is not implemented,
so it doesn't know that the name is already
in UNC. So I've found no better way that to
just duplicate the UNC printer name into
`drives[dd].root`. Even now I don't see how
this can be avoided w/o implementing
`QUALIFY_FILENAME` (which we decided
not to do because of the too small buffers).